src/randomenv.cpp

```diff
@@ -0,0 +1,389 @@
+template<typename T>
+CSHA512& operator<<(CSHA512& hasher, const T& data) {
+    hasher.Write((const unsigned char*)&data, sizeof(data));
+    return hasher;
+}
+
+#ifndef WIN32
+void AddSockaddr(CSHA512& hasher, const struct sockaddr *addr)
+{
+    if (addr == nullptr) return;
+    switch (addr->sa_family) {
+    case AF_INET:
+        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in));
+        break;
+    case AF_INET6:
+        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in6));
+        break;
+    default:
+        hasher.Write((const unsigned char*)&addr->sa_family, sizeof(addr->sa_family));
+    }
+}
+
+void AddStat(CSHA512& hasher, const struct stat& sb)
+{
+    hasher << sb.st_dev << sb.st_ino << sb.st_mode << sb.st_nlink << sb.st_uid << sb.st_gid << sb.st_size << sb.st_blksize << sb.st_blocks << sb.st_atime << sb.st_mtime << sb.st_ctime;
+}
+
+void AddFile(CSHA512& hasher, const char *path)
+{
+    struct stat sb;
+    int f = open(path, O_RDONLY);
+    if (f != -1) {
+        unsigned char fbuf[1024];
+        int n;
+        hasher.Write((const unsigned char*)&f, sizeof(f));
+        if (fstat(f, &sb) == 0) {
+            AddStat(hasher, sb);
+        }
+        do {
+            n = read(f, fbuf, sizeof(fbuf));
+            if (n > 0) hasher.Write(fbuf, n);
+            /* not bothering with EINTR handling. */
+        } while (n == sizeof(fbuf));
+        close(f);
+    }
+}
+
+void AddPath(CSHA512& hasher, const char *path)
+{
+    struct stat sb;
+    if (stat(path, &sb) == 0) {
+        hasher.Write((const unsigned char*)path, strlen(path) + 1);
+        AddStat(hasher, sb);
+    }
+}
+#endif
+
+void AddSysctl(CSHA512& hasher, const char* path)
+{
+#if HAVE_DECL_SYSCTLBYNAME
+    unsigned char buffer[4096];
+    size_t siz = 4096;
+    int ret = sysctlbyname(path, buffer, &siz, nullptr, 0);
+    if (ret == 0 || (ret == -1 && errno == ENOMEM)) {
+        hasher.Write((const unsigned char*)path, strlen(path) + 1);
+        hasher.Write(buffer, std::min<size_t>(siz, 4096));
+    }
+#endif
+}
+
+#ifdef HAVE_GETCPUID
+void AddCPUID(CSHA512& hasher)
+{
+    uint32_t ax, bx, cx, dx;
+    GetCPUID(0, 0, ax, bx, cx, dx);
+    hasher << ax << bx << cx << dx;
+    uint32_t max = ax;
+    for (uint32_t leaf = 1; leaf <= max; ++leaf) {
+        for (uint32_t subleaf = 0;; ++subleaf) {
+            GetCPUID(leaf, subleaf, ax, bx, cx, dx);
+            hasher << leaf << subleaf << ax << bx << cx << dx;
+            // Iterate over subleaves for leaf 4, 11, 13
+            if (leaf != 4 && leaf != 11 && leaf != 13) break;
+            if ((leaf == 4 || leaf == 13) && ax == 0) break;
+            if (leaf == 11 && (cx & 0xFF00) == 0) break;
+        }
+    }
+}
+#endif
+} // namespace
+
+void RandAddDynamicEnv(CSHA512& hasher)
+{
+    RandAddSeedPerfmon(hasher);
+
+#ifdef WIN32
+    FILETIME ftime;
+    GetSystemTimeAsFileTime(&ftime);
+    hasher << ftime;
+#else
+#  ifdef __MACH__
+    hasher << mach_absolute_time();
```

Comment:
There's not a heap of documentation on  `mach_absolute_time` and I can't find any calls to it in recent OpenSSL, however looks like it's available for macOS versions > 10.9. `/usr/include/mach/mach_time.h`.

Reply:
Yeah, I tried to find an equivalent to CLOCK_MONOTONIC on MacOS, and found this. It doesn't exactly match what OpenSSL did.